### PR TITLE
[CARBONDATA-2553] support ZSTD compression for sort temp file

### PR DESCRIPTION
**Pull request by kevinjmh**
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 

ZSTD project Home： http://facebook.github.io/zstd/

Zstandard is a real-time compression algorithm, providing high compression ratios. It can offer higher compression  ratio than snappy with similar speed. In our test, we find that the file size of sort temp file generated by zstd almost half of  snappy. 

Hadoop and Spark also imported this project. See:

https://github.com/apache/hadoop/commit/a0a276162147e843a5a4e028abdca5b66f5118da

https://github.com/apache/spark/commit/444bce1c98c45147fe63e2132e9743a0c5e49598

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5129/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6145/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4982/

**Comment by jackylk** [comment]
Can you provide description for this PR? And introduce the ZSTD library?

**Comment by kevinjmh** [comment]
PR description  update

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4984/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java [hunk 1]
```diff
@@ -1290,11 +1290,11 @@ public String getSortTempCompressor() {
     String compressor = getProperty(CarbonCommonConstants.CARBON_SORT_TEMP_COMPRESSOR,
         CarbonCommonConstants.CARBON_SORT_TEMP_COMPRESSOR_DEFAULT).toUpperCase();
     if (compressor.isEmpty() || "SNAPPY".equals(compressor) || "GZIP".equals(compressor)
-        || "BZIP2".equals(compressor) || "LZ4".equals(compressor)) {
+        || "BZIP2".equals(compressor) || "LZ4".equals(compressor) || "ZSTD".equals(compressor)) {
       return compressor;
     } else {
       LOGGER.warn("The ".concat(CarbonCommonConstants.CARBON_SORT_TEMP_COMPRESSOR)
-          .concat(" configuration value is invalid. Only snappy,gzip,bip2,lz4 and")
+          .concat(" configuration value is invalid. Only snappy,gzip,bip2,lz4,zstd and")
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java]: change to `, zstd`

File: core/src/main/java/org/apache/carbondata/core/datastore/filesystem/LocalCarbonFile.java [hunk 2]
```diff
@@ -368,6 +372,8 @@ public DataOutputStream getDataOutputStream(String path, FileFactory.FileType fi
       outputStream = new SnappyOutputStream(new FileOutputStream(path));
     } else if ("LZ4".equalsIgnoreCase(compressor)) {
       outputStream = new LZ4BlockOutputStream(new FileOutputStream(path));
+    } else if ("ZSTD".equalsIgnoreCase(compressor)) {
+      outputStream = new ZstdOutputStream(new FileOutputStream(path), 1);
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/filesystem/LocalCarbonFile.java]: what is the meaning of 1?

Comment (kevinjmh) [reply core/src/main/java/org/apache/carbondata/core/datastore/filesystem/LocalCarbonFile.java]: compress level 1.  We have test using default level 3 which can get better compress ratio but cost more time. As for sort temp file which is not used for storage, level 1 is cost-effective.

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6155/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4994/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5137/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6181/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5020/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6323/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5160/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6357/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5195/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**